codec-http2/src/main/java/io/netty/handler/codec/http2/CleartextHttp2ServerUpgradeHandler.java

```diff
@@ -64,38 +63,27 @@ public CleartextHttp2ServerUpgradeHandler(HttpServerCodec httpServerCodec,
     }
 
     @Override
-    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
-        ctx.pipeline()
-           .addBefore(ctx.name(), null, new PriorKnowledgeHandler())
-           .addBefore(ctx.name(), null, httpServerCodec)
-           .replace(this, null, httpServerUpgradeHandler);
-    }
-
-    /**
-     * Peek inbound message to determine current connection wants to start HTTP/2
-     * by HTTP upgrade or prior knowledge
-     */
-    private final class PriorKnowledgeHandler extends ByteToMessageDecoder {
-        @Override
-        protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
-            int prefaceLength = CONNECTION_PREFACE.readableBytes();
-            int bytesRead = Math.min(in.readableBytes(), prefaceLength);
-
-            if (!ByteBufUtil.equals(CONNECTION_PREFACE, CONNECTION_PREFACE.readerIndex(),
-                                    in, in.readerIndex(), bytesRead)) {
-                ctx.pipeline().remove(this);
-            } else if (bytesRead == prefaceLength) {
-                // Full h2 preface match, removed source codec, using http2 codec to handle
-                // following network traffic
-                ctx.pipeline()
-                   .remove(httpServerCodec)
-                   .remove(httpServerUpgradeHandler);
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
+        int prefaceLength = CONNECTION_PREFACE.readableBytes();
+        int bytesRead = Math.min(in.readableBytes(), prefaceLength);
+        //
+        // Peek inbound message to determine current connection wants to start HTTP/2
+        // by HTTP upgrade or prior knowledge
+        if (!ByteBufUtil.equals(CONNECTION_PREFACE, CONNECTION_PREFACE.readerIndex(),
+                in, in.readerIndex(), bytesRead)) {
 
-                ctx.pipeline().addAfter(ctx.name(), null, http2ServerHandler);
-                ctx.pipeline().remove(this);
+            // no prior knowledge used.
+            ctx.pipeline()
+                    .addAfter(ctx.name(), null, httpServerUpgradeHandler)
+                    .addAfter(ctx.name(), null, httpServerCodec);
 
```

Comment:
This is still going to have a weird usage pattern in that if you want to have handlers dealing with HTTP messages the pipeline is going to go from this:
```
[CleartextHttp2ServerUpgradeHandler, SeverUpgradeHandler, ServerCodec, MyHttpMessageHandler(s)]
```
to
```
[SeverUpgradeHandler, MyHttpMessageHandler(s)]
```
The better thing to do would be to inject the right channel initializer for either a H2C or cleartext channel.

Reply:
@bryce-anderson while I agree I think it is nothing we can do in 4.1.x easily. Thats why I opted for this which still gives a little improvement.  WDYT ?

Reply:
Just my opinion, but I like the collapsing of the `CleartextHttp2ServerUpgradeHandler` into the `ByteToMessageDecoder` and we should do that for sure. Wrt the pipeline manipulation: it's ugly on 4.1 and it will be ugly after this patch _plus_ require people to change how their pipelines are setup. I suspect we can do better in netty5. That said, I'm not strongly opinionated about it so don't let my opinion stand in the way of progress if you feel this is much better.

Reply:
@bryce-anderson I wonder why you think people require to change how they setup the pipeline after this PR. Can you explain ?

Reply:
@bryce-anderson ping

Reply:
@bryce-anderson I did factor our the "merge" into another PR as it seems this is a no-brainer for you . https://github.com/netty/netty/pull/10008